spcl/dace
# Add Support for HBM and ND-copies on FPGA (pull request 2)

Changes:

sdfg/utils: 
- Added several utility functions for HBM-arrays/subsets to utils

targets/cpp: 
- Modifed ptr to generate distinct names for HBM-arrays. The function array_interface_variable is only called from ptr.
- Modified several functions in targets/cpp slighly to support HBM-arrays
- targets/cpp: added function to_cpp_arr, which generates C++ arrays
- targets/cpp: Split ndcopy_to_strided_copy into 2 functions, is_1d_nostrided_copy to check if a copy is 1d and ndcopy_to_strided_copy which behaves as before (but uses is_1d_nostrided_copy internally)

targets/cpu:
- Modifed calls to cpp.ptr in some locations
- Added NotImplementedException if nested sdfg's on cpu side deal with HBM-arrays. This will likely be supported in the next PR, but at the moment it seems to much like a nice to have.

targets/fpga:
- Modifed several of the functions which generate arguments/definitions and so on for variables to create multiple of those objects in case of HBM-arrays
- make_parameters recognizes and passes HBM and DDR in it's bank_assignments return value
- emit_copy has ndim (up to 3D) block copies. The subsets dimensions are reverted when compared to numpy, which maybe should be changed depending on the conventions of dace
- emit_copy has support for Host2Device, Device2Host, Device2Device and on Device copies from/to HBM-arrays and also between HBM and DDR. Writing directly from HBM to Streams is not yet supported (it can be done using tasklets with single bank memlets of course)
- copy_memory "disolves" the magic index for HBM-multibank memlets and only calls emit_copy with single bank HBM memlets
- found a possible old bug (maybe this is untested) at line 1423. Altered it to what seems to make sense there.

targets/xilinx:
- Modifed several of the functions which generate arguments/definitions and so on for variables to create multiple of those objects in case of HBM-arrays
- get_generated_codeobjects: HBM interfaces are declared

sdfg/validation:
- Check that no HBM multibank memlets are attached to a tasklet
- Check that the bank/hbmbank location property is valid for the given array (correct type, correct arrayshape, ...)
- Check that all HBM-memlets have constant evaluatable indices

transformation/fpga_transform_state
- A few smaller modifications so that hbmbank is correctly set

tests/fpga/*
- Added several tests for all the mentioned changes above
- Added hbm_hw_emu_runner a .sh file to run all the hbm tests as hw_emu, since emulation seems to run into trouble when programming it with multiple .xclbins

dace/data/*
- Added the value type of the locationproperty to str from pystr_to_symbolic. Needed because of the serialization for ranges
